### PR TITLE
Fix #2221, DenseNet issue with gradient checkpoints (#2236)

### LINKED ISSUE #2221: DenseNet memory_efficient mode appears subtly broken since torchscript changes
## 🐛 Bug

I was working with DenseNet recently, I noticed memory_efficient training was faster than with the flag disabled (unexpected), and it wasn't training very well at all, also errors from checkpoint fn about args not having required_grad=True set.

I believe #1342 subtly broke the checkpointing by wraping inputs in an extra layer of container

densenet.py:53
``` 
    def call_checkpoint_bottleneck(self, input):
        # type: (List[Tensor]) -> Tensor
        def closure(*inputs):
            return self.bn_function(*inputs)

        return cp.checkpoint(closure, input)
```

should be

```
    def call_checkpoint_bottleneck(self, input):
        # type: (List[Tensor]) -> Tensor
        def closure(*inputs):
            return self.bn_function(inputs)  # remove *

        return cp.checkpoint(closure, *input)  # add *
```
## To Reproduce

Train any densenet network with `memory_efficient=True`

## Environment

Any

 - PyTorch / torchvision Version (e.g., 1.0 / 0.4.0): 1.4
 - OS (e.g., Linux): any

### PR DESCRIPTION
**Pull request by fmassa**
* Fix #2221, DenseNet issue with gradient checkpoints (memory_efficient=True)

* Add grad/param count test for mem_efficient densenet

### CODE REVIEW

**Comment by fmassa** [comment]
Merged in 9627194
